File: ctci/chapter_1_arrays_strings/js/1.5-oneAway.js
/* O(n) solution */

/**
 * Compare two strings to determine if they are one edit away from each
 * other.
 * @param   {string}  str1
 * @param   {string}  str2
 * @returns {boolean}
 */
const isOneEditAway = (str1, str2) => {
  const len1 = str1.length
  const len2 = str2.length
  if (len1 === len2) return isOneEditReplace(str1, str2)
  if (Math.abs(len1 - len2) === 1) return isOneEditInsert(str2, str1)
  return false
}

/**
 * Counts the number of differences between two strings of same length.
 * Returns true if the difference is at most one off.
 * @param   {string}  str1
 * @param   {string}  str2
 * @returns {boolean}
 */
const isOneEditReplace = (str1, str2) => {
  let differences = 0
  for (const i in str1) if (str1[i] !== str2[i]) differences++
  return differences <= 1
}

/**
 * Determines shorter string and loops from start to finish. If there is a
 * difference, longer string is adjusted. If there is more than one
 * difference, returns false.
 * @param   {string}  str1
 * @param   {string}  str2
 * @returns {boolean}
 */
const isOneEditInsert = (str1, str2) => {
  const short = str1.length > str2.length ? str2 : str1
  const long = str2.length > str1.length ? str2 : str1
  let foundDifference = false
  for (let i = 0, j = 0; i < short.length; i++, j++) {
    if (short[i] !== long[j]) {
      if (foundDifference) return false
      foundDifference = true
      j++
    }
  }
  return true
}

console.log(isOneEditAway("beal", "bal")) // true
console.log(isOneEditAway("val", "bal")) // true
console.log(isOneEditAway("val", "bale")) // false
console.log(isOneEditAway("mail", "mailman")) // false
console.log(isOneEditAway("abc", "abcde")) // false
console.log(isOneEditAway("abc", "abcd")) // true

File: ctci/chapter_1_arrays_strings/1.5-one_away.py
# O(n) solution


def is_one_edit_replace(str1: str, str2: str) -> bool:
    """
    Counts the number of differences between two strings of same length.
    Returns true if the difference is at most one off.
    """
    differences = 0
    for i in range(len(str1)):
        if str1[i] != str2[i]:
            differences += 1
    return differences <= 1


def is_one_edit_insert(str1: str, str2: str) -> bool:
    """
    Determines shorter string and loops from start to finish. If there is a
    difference, longer string is adjusted. If there is more than one
    difference, returns false.
    """
    if len(str1) > len(str2):
        long, short = str1, str2
    else:
        short, long = str1, str2
    i = j = 0
    found_difference = False
    while i < len(short):
        if (short[i] != long[j]):
            if found_difference:
                return False
            found_difference = True
            j += 1
        i += 1
        j += 1
    return True


def is_one_edit_away(str1: str, str2: str) -> bool:
    """
    Compare two strings to determine if they are one edit away from each
    other.
    """
    len1 = len(str1)
    len2 = len(str2)
    if len1 == len2:
        return is_one_edit_replace(str1, str2)
    if abs(len1 - len2) == 1:
        return is_one_edit_insert(str1, str2)
    return False


if __name__ == 'main':
    print(is_one_edit_away('beal', 'bal'))  # True
    print(is_one_edit_away('val', 'bal'))  # True
    print(is_one_edit_away('val', 'bale'))  # False
    print(is_one_edit_away('mail', 'mailman'))  # False
    print(is_one_edit_away('abc', 'abcde'))  # False
    print(is_one_edit_away('', 'a'))  # True
